### PR TITLE
Config: RegisterObserver init config.

### PR DESCRIPTION
**Pull request by eloycoto**
When yggd restart and any workload is already running, if the config
didn't change, it will not update the config to all registered observers.

With this commit, when an observer is created, has the option to send the
initial config to the Update() method, so the observers can run the initial
config and work against it.

Signed-off-by: Eloy Coto <eloy.coto@acalustra.com>

### CODE REVIEW

File: cmd/device-worker/main.go [hunk 1]
```diff
@@ -100,32 +100,32 @@ func main() {
...
-	configManager.RegisterObserver(wl)
+	configManager.RegisterObserver(wl, true)
 	wl.RegisterObserver(workloadMetricWatcher)
 
 	hw := hardware2.Hardware{}
 
 	dataMonitor := datatransfer.NewMonitor(wl, configManager)
 	wl.RegisterObserver(dataMonitor)
-	configManager.RegisterObserver(dataMonitor)
+	configManager.RegisterObserver(dataMonitor, false)
```
Comment (eloycoto) [inline cmd/device-worker/main.go:120]: @jakub-dzon  about this one, I'm not sure if we should wait until get the new s3 config, it's not clear to me.

**Comment by ydayagi** [comment]
I think it is not so good for us to mix registration and update. The whole register mechanism becomes very complicated. My personal opinion is that observers are not needed. But I guess we're not going to change that. What I suggest (if possible) is to add a step after registering all observers that calls Update explicitly. Or add a step after registering the observers that calls observers' Update explicitly. Only for the observers that require it, of course.

File: internal/configuration/configuration.go [hunk 2]
```diff
@@ -69,7 +69,13 @@ func NewConfigurationManager(dataDir string) *Manager {
 	return &mgr
 }
 
-func (m *Manager) RegisterObserver(observer Observer) {
+// RegisterObserver registers an Observer to the config, that will get called
+// when any config changes happens.
+// If withInitConfig is set, observer will be called with the current config.
+func (m *Manager) RegisterObserver(observer Observer, withInitConfig bool) {
```
Comment (jakub-dzon) [inline internal/configuration/configuration.go:75]: I think managing dependencies between observers might be easier if the initial Update round would be performed at the same time for all of them, after they have been registered.
So the flow would be:
```
configManager.RegisterObserver(a, true)
configManager.RegisterObserver(b, false)
configManager.RegisterObserver(c, true)
// ...
configManager.Update() // or configManager.NotifyObservers() that would call Update on each observier with initialization enabled
```

File: cmd/device-worker/main.go (hunk 1 above)
Comment (jakub-dzon) [reply cmd/device-worker/main.go:120]: Right now it should be `false`, but we should adjust `Monitor.Start()` to not use `ForceUpdate` and set it to `true` then

File: internal/configuration/configuration.go (hunk 2 above)
Comment (eloycoto) [reply internal/configuration/configuration.go:75]: It'll not work, mainly because when workload start, emit some WorkloadStarted messages, and metrics does not have the config, etc.. It's a bit of chicken-egg issue.

File: cmd/device-worker/main.go [hunk 3]
```diff
@@ -100,32 +100,32 @@ func main() {
 
 	// --- Client metrics configuration ---
 	metricsStore, err := metrics.NewTSDB(dataDir)
-	configManager.RegisterObserver(metricsStore)
+	configManager.RegisterObserver(metricsStore, true)
```
Comment (jakub-dzon) [inline cmd/device-worker/main.go:103]: I will fix `TSDB` later not to be intialized with defaults in the factory method, but on the first update.

File: internal/configuration/configuration.go (hunk 2 above)
Comment (jakub-dzon) [reply internal/configuration/configuration.go:75]: Understood. :+1:

**Comment by eloycoto** [comment]
> I think it is not so good for us to mix registration and update. The whole register mechanism becomes very complicated. My personal opinion is that observers are not needed. But I guess we're not going to change that. What I suggest (if possible) is to add a step after registering all observers that calls Update explicitly. Or add a step after registering the observers that calls observers' Update explicitly. Only for the observers that require it, of course.

@ydayagi  is this a hard request? could we defer to the next iteration after the migration on the project?

File: internal/configuration/configuration.go [hunk 4]
```diff
@@ -69,7 +69,13 @@ func NewConfigurationManager(dataDir string) *Manager {
...
+	if withInitConfig {
```
Comment (masayag) [inline internal/configuration/configuration.go:76]: I was thinking on a a slightly different way to implement this:
The `Observer` interface can be extended with `Init` method that will be responsible to making sure the observer is set properly (it will be each observer responsibility if to call Update from Init or perform any initialization logic or do nothing).
```
type Observer interface {
	Init(configuration models.DeviceConfigurationMessage) error
	Update(configuration models.DeviceConfigurationMessage) error
}
```
Then the main will remain mostly the same, without a need to specify for each observer how it should be started.
In other words, the responsibility to init the observer will be done within each observer.

Comment (eloycoto) [reply internal/configuration/configuration.go:76]: Makes a lot of sense to me, I'll do in this way. Thanks @masayag

File: internal/metrics/system.go [hunk 5]
```diff
@@ -39,6 +39,10 @@ func NewSystemMetrics(daemon MetricsDaemon, config DeviceConfigurationProvider)
 	return sm
 }
 
+func (sm *SystemMetrics) Init(config models.DeviceConfigurationMessage) error {
+	return sm.Update(config)
```
Comment (jakub-dzon) [inline internal/metrics/system.go:43]: Initialization that is performed in `NewSystemMetrics` should be removed.

File: internal/datatransfer/monitor.go [hunk 6]
```diff
@@ -41,9 +41,6 @@ func NewMonitor(workloadsManager *workload.WorkloadManager, configManager *confi
 func (m *Monitor) Start() {
 	// Made this to force the config load in case there is no storage defined yet
```
Comment (jakub-dzon) [inline internal/datatransfer/monitor.go:42]: That comment should be removed

**Review by jakub-dzon — CHANGES_REQUESTED** [review]

File: internal/datatransfer/monitor.go (hunk 6 above)
Comment (eloycoto) [reply internal/datatransfer/monitor.go:42]: :facepalm: sorry

**Review by pkliczewski — COMMENTED** [review]
/lgtm

**Review by masayag — APPROVED** [review]

**Review by jakub-dzon — APPROVED** [review]
